server/src/main/java/com/genymobile/scrcpy/MediaCodecCompat.java

```diff
@@ -0,0 +1,129 @@
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Version of {@link MediaCodec} that backports {@link #getOutputBuffer} to Kitkat.
+ * The backported implementation isn't thread safe.
+ */
+abstract class MediaCodecCompat {
+    private final MediaCodec delegate;
+
+    protected MediaCodecCompat(@NonNull MediaCodec delegate) {
+        this.delegate = delegate;
+    }
+
+    @NonNull
+    protected MediaCodec getDelegate() {
+        return delegate;
+    }
+
+    @NonNull
+    static MediaCodecCompat wrap(@NonNull MediaCodec codec) {
+        if (SDK_INT >= 21) {
+            return new Platform(codec);
+        } else {
+            return new Backport(codec);
+        }
+    }
+
+    abstract int dequeueOutputBuffer(
+            @NonNull MediaCodec.BufferInfo info, long timeoutUs);
+
+    @Nullable
+    abstract ByteBuffer getOutputBuffer(int index);
+
+    abstract void releaseOutputBuffer(int index, boolean render);
+
+    @SuppressWarnings("deprecation")
+    private static class Backport extends MediaCodecCompat {
+        private ByteBuffer[] cachedOutputBuffers = null;
+
+        Backport(@NonNull MediaCodec delegate) {
+            super(delegate);
+        }
+
+        @Override
+        int dequeueOutputBuffer(
+                @NonNull MediaCodec.BufferInfo info, long timeoutUs) {
+            final int res = getDelegate().dequeueOutputBuffer(info, timeoutUs);
+            if (res == INFO_OUTPUT_BUFFERS_CHANGED) {
+                cachedOutputBuffers = null;
+            } else if (res >= 0) {
+                validateOutputByteBuffer(cachedOutputBuffers, res, info);
+            }
+            return res;
+        }
+
+        @Nullable
+        @Override
+        ByteBuffer getOutputBuffer(int index) {
+            if (cachedOutputBuffers == null) {
+                cacheOutputBuffers();
+            }
+            if (cachedOutputBuffers == null) {
+                throw new IllegalStateException();
+            }
+            return cachedOutputBuffers[index];
+        }
+
+        @Override
+        void releaseOutputBuffer(int index, boolean render) {
+            cachedOutputBuffers = null;
+            getDelegate().releaseOutputBuffer(index, render);
+        }
+
+        private void cacheOutputBuffers() {
+            ByteBuffer[] buffers = null;
+            try {
+                buffers = getDelegate().getOutputBuffers();
+                // don't actually invalidate the buffer as it's private platform API
+            } catch (IllegalStateException e) {
+                // we don't get buffers in async mode
+            }
+            cachedOutputBuffers = buffers;
+        }
+
+        private void validateOutputByteBuffer(
+                @Nullable ByteBuffer[] buffers, int index, @NonNull MediaCodec.BufferInfo info) {
+            if (buffers != null && index >= 0 && index < buffers.length) {
```

Comment:
Come to think about it, this may be necessary only for "invalidated" buffers which had position/limit reset. The hidden API isn't even available on Kitkat. I'll remove this method entirely. Thanks!